### ISSUE TITLE
Spacy NER model Error : [E103] Trying to set conflicting doc.ents: '(1417, 1423, 'Companies worked at')' and '(1356, 1793, 'Skills')'. A token can only be part of one entity, so make sure the entities you're setting don't overlap.

### ISSUE THREAD

**Issue by saaddine** [explosion/spaCy#5240]
## How to reproduce the behaviour

import json
import random
import logging
from sklearn.metrics import classification_report
from sklearn.metrics import precision_recall_fscore_support
from spacy.gold import GoldParse
from spacy.scorer import Scorer
from sklearn.metrics import accuracy_score
def convert_dataturks_to_spacy(dataturks_JSON_FilePath):
    try:
        training_data = []
        lines=[]
        with open(dataturks_JSON_FilePath, 'r',encoding="utf8") as f:
            lines = f.readlines()

        for line in lines:
            data = json.loads(line)
            text = data['content']
            entities = []
            for annotation in data['annotation']:
                #only a single point in text annotation.
                point = annotation['points'][0]
                labels = annotation['label']
                # handle both list of labels or a single label.
                if not isinstance(labels, list):
                    labels = [labels]

                for label in labels:
                    #dataturks indices are both inclusive [start, end] but spacy is not [start, end)
                    entities.append((point['start'], point['end'] + 1 ,label))

            training_data.append((text, {"entities" : entities}))

        return training_data
    except Exception as e:
        logging.exception("Unable to process " + dataturks_JSON_FilePath + "\n" + "error = " + str(e))
        return None

import spacy
################### Train Spacy NER.###########
def train_spacy():

    TRAIN_DATA = convert_dataturks_to_spacy("C:/Users/LENOVO/Desktop/PFE/CRF/Entity-Recognition-In-Resumes-SpaCy/traindata.json")
    nlp = spacy.blank('en')  # create blank Language class
    # create the built-in pipeline components and add them to the pipeline
    # nlp.create_pipe works for built-ins that are registered with spaCy
    if 'ner' not in nlp.pipe_names:
        ner = nlp.create_pipe('ner')
        nlp.add_pipe(ner, last=True)
       

    # add labels
    for _, annotations in TRAIN_DATA:
         for ent in annotations.get('entities'):
            ner.add_label(ent[2])

    # get names of other pipes to disable them during training
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
    with nlp.disable_pipes(*other_pipes):  # only train NER
        optimizer = nlp.begin_training()
        for itn in range(10):
            print("Statring iteration " + str(itn))
            random.shuffle(TRAIN_DATA)
            losses = {}
            for text, annotations in TRAIN_DATA:
                nlp.update(
                    [text],  # batch of texts
                    [annotations],  # batch of annotations
                    drop=0.2,  # dropout - make it harder to memorise data
                    sgd=optimizer,  # callable to update weights
                    losses=losses)
            print(losses)
    #test the model and evaluate it
    examples = convert_dataturks_to_spacy("C:/Users/LENOVO/Desktop/PFE/CRF/Entity-Recognition-In-Resumes-SpaCy/testdata.json")
    tp=0
    tr=0
    tf=0

    ta=0
    c=0        
    for text,annot in examples:

        f=open("resume"+str(c)+".txt","w")
        doc_to_test=nlp(text)
        d={}
        for ent in doc_to_test.ents:
            d[ent.label_]=[]
        for ent in doc_to_test.ents:
            d[ent.label_].append(ent.text)

        for i in set(d.keys()):

            f.write("\n\n")
            f.write(i +":"+"\n")
            for j in set(d[i]):
                f.write(j.replace('\n','')+"\n")
        d={}
        for ent in doc_to_test.ents:
            d[ent.label_]=[0,0,0,0,0,0]
        for ent in doc_to_test.ents:
            doc_gold_text= nlp.make_doc(text)
            gold = GoldParse(doc_gold_text, entities=annot.get("entities"))
            y_true = [ent.label_ if ent.label_ in x else 'Not '+ent.label_ for x in gold.ner]
            y_pred = [x.ent_type_ if x.ent_type_ ==ent.label_ else 'Not '+ent.label_ for x in doc_to_test]  
            if(d[ent.label_][0]==0):
                #f.write("For Entity "+ent.label_+"\n")   
                #f.write(classification_report(y_true, y_pred)+"\n")
                (p,r,f,s)= precision_recall_fscore_support(y_true,y_pred,average='weighted')
                a=accuracy_score(y_true,y_pred)
                d[ent.label_][0]=1
                d[ent.label_][1]+=p
                d[ent.label_][2]+=r
                d[ent.label_][3]+=f
                d[ent.label_][4]+=a
                d[ent.label_][5]+=1
        c+=1
    for i in d:
        print("\n For Entity "+i+"\n")
        print("Accuracy : "+str((d[i][4]/d[i][5])*100)+"%")
        print("Precision : "+str(d[i][1]/d[i][5]))
        print("Recall : "+str(d[i][2]/d[i][5]))
        print("F-score : "+str(d[i][3]/d[i][5]))
train_spacy()
 

## Your Environment

* Operating System:
* Python Version Used:3.7.4
* spaCy Version Used: 2.2.4
* Environment Information:
Python 3
Spyder 
Windows 10

**Comment by saaddine** [comment]
i'm using the annotated training set  : [https://dataturks.com/projects/abhishek.narayanan/Entity%20Recognition%20in%20Resumes](url)

**Comment by svlandeg** [comment]
The error message is already quite descriptive: 
> A token can only be part of one entity, so make sure the entities you're setting don't overlap.

spaCy's NER annotations are assumed to be non-overlapping, so you'll have to ensure your training data adheres to that principle before feeding it into the pipeline.

**Comment by saaddine** [comment]
but i think that someone used the same dataset and it's okay for him  . i mean is it possble to pass a flag or something that the code override the overlapping because the dataset is tooo big and impossible to eliminate overlapping

**Comment by svlandeg** [comment]
I'm afraid it simply isn't possible. The current architecture of spaCy only allows one entity per token - there is no workaround for now other than to filter the input dataset.

One thing you could do is to check which type of entities usually overlap. In your original example, the label "Companies worked at" clashes with "Companies worked at". It might be that one of these types/labels is less relevant to your use-case; then you can remove those annotations to get non-overlapping input data.

The other option is to create separate datasets: e.g. one for Skills and one for "Companies worked at", and train two NER models separately. But at prediction time you'll still have to implement some logic to combine the two, because if they each end up predicting an entity for the same token, you won't be able to retain both, you'll have to pick one for that token.

**Comment by svlandeg** [comment]
PS: I personally find it a little odd that there are tokens annotated as part of an entity "Company" as well as "Skill". I also note that the "Skills" entity is 400 tokens long - that seems excessive?
